File: web/src/sim/diplomacy/Relation.ts
export enum Relation {
  Peace = 0,
  RightOfWay = 1,
  SharedIntelligence = 2,
  Ceasefire = 3,
  War = 4,
}

/*
 * Hubungan bersifat simetris dan disimpan di segitiga bawah sebuah matriks,
 * yang menjaga dunia 247 negara tetap 30 KB alih-alih 60 dan membuat dua negara
 * mustahil berbeda pendapat soal apakah mereka sedang berperang.
 */
export class Relations {
  readonly nationCount: number

  readonly #matrix: Uint8Array

  constructor(nationCount: number) {
    if (nationCount < 0) {
      throw new RangeError(`jumlah negara tidak boleh negatif: ${nationCount}`)
    }

    this.nationCount = nationCount
    this.#matrix = new Uint8Array((nationCount * (nationCount - 1)) / 2)
  }

  between(a: number, b: number): Relation {
    if (a === b) return Relation.Peace

    return this.#matrix[this.#indexOf(a, b)]! as Relation
  }

  set(a: number, b: number, relation: Relation): void {
    if (a === b) {
      throw new RangeError('sebuah negara tidak bisa berhubungan dengan dirinya sendiri')
    }

    this.#matrix[this.#indexOf(a, b)] = relation
  }

  atWar(a: number, b: number): boolean {
    return this.between(a, b) === Relation.War
  }

  mayEnter(mover: number, owner: number): boolean {
    if (mover === owner) return true

    const relation = this.between(mover, owner)
    return relation === Relation.War || relation === Relation.RightOfWay
  }

  /* Larik, bukan generator: urutannya adalah urutan indeks negara dan pemanggil
     di simulasi mengulanginya lebih dari sekali per hari. */
  enemiesOf(nation: number): number[] {
    const enemies: number[] = []

    for (let other = 0; other < this.nationCount; other++) {
      if (other !== nation && this.atWar(nation, other)) enemies.push(other)
    }

    return enemies
  }

  #indexOf(a: number, b: number): number {
    if (a < 0 || a >= this.nationCount || b < 0 || b >= this.nationCount) {
      throw new RangeError(`negara ${a} atau ${b} di luar 0..${this.nationCount - 1}`)
    }

    const high = a > b ? a : b
    const low = a > b ? b : a

    return (high * (high - 1)) / 2 + low
  }
}

File: sim/NationRise.Core/Diplomacy/Relation.cs
namespace NationRise.Core.Diplomacy;

public enum Relation : byte
{
    Peace = 0,
    RightOfWay = 1,
    SharedIntelligence = 2,
    Ceasefire = 3,
    War = 4,
}

/*
   Relations are symmetric and stored in the lower triangle of a matrix, which
   keeps a 247-nation world at 30 KB instead of 60 and makes it impossible for
   two nations to disagree about whether they are at war.
*/
public sealed class Relations
{
    private readonly Relation[] _matrix;

    public int NationCount { get; }

    public Relations(int nationCount)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(nationCount);
        NationCount = nationCount;
        _matrix = new Relation[nationCount * (nationCount - 1) / 2];
    }

    public Relation Between(int a, int b)
    {
        if (a == b)
        {
            return Relation.Peace;
        }

        return _matrix[IndexOf(a, b)];
    }

    public void Set(int a, int b, Relation relation)
    {
        if (a == b)
        {
            throw new ArgumentException("A nation cannot have a relation with itself.");
        }

        _matrix[IndexOf(a, b)] = relation;
    }

    public bool AtWar(int a, int b) => Between(a, b) == Relation.War;

    public bool MayEnter(int mover, int owner) =>
        mover == owner || Between(mover, owner) is Relation.War or Relation.RightOfWay;

    public IEnumerable<int> EnemiesOf(int nation)
    {
        for (int other = 0; other < NationCount; other++)
        {
            if (other != nation && AtWar(nation, other))
            {
                yield return other;
            }
        }
    }

    private int IndexOf(int a, int b)
    {
        if ((uint)a >= (uint)NationCount || (uint)b >= (uint)NationCount)
        {
            throw new ArgumentOutOfRangeException(nameof(a));
        }

        (int high, int low) = a > b ? (a, b) : (b, a);
        return high * (high - 1) / 2 + low;
    }
}
